cosmos/cache.py

```diff
@@ -0,0 +1,124 @@
+# This can be reviewed in the future.
+def create_cache_identifier(dag: DAG, task_group: TaskGroup | None) -> str:
+    """
+    Given a DAG name and a (optional) task_group_name, create the identifier for caching.
+
+    :param dag_name: Name of the Cosmos DbtDag being cached
+    :param task_group_name: (optional) Name of the Cosmos DbtTaskGroup being cached
+    :return: Unique identifier representing the cache
+    """
+    if task_group:
+        if task_group.dag_id is not None:
+            cache_identifiers_list = [task_group.dag_id]
+        if task_group.group_id is not None:
+            cache_identifiers_list.extend([task_group.group_id.replace(".", "_")])
+        cache_identifier = "_".join(cache_identifiers_list)
+    else:
+        cache_identifier = dag.dag_id
+
+    return cache_identifier
+
+
+def obtain_cache_dir_path(cache_identifier: str, base_dir: Path = settings.cache_dir) -> Path:
+    """
+    Return a directory used to cache a specific Cosmos DbtDag or DbtTaskGroup. If the directory
+    does not exist, create it.
+
+    :param cache_identifier: Unique key used as a cache identifier
+    :param base_dir: Root directory where cache will be stored
+    :return: Path to directory used to cache this specific Cosmos DbtDag or DbtTaskGroup
+    """
+    cache_dir_path = base_dir / cache_identifier
+    tmp_target_dir = cache_dir_path / DBT_TARGET_DIR_NAME
+    tmp_target_dir.mkdir(parents=True, exist_ok=True)
+    return cache_dir_path
+
+
+def get_timestamp(path: Path) -> float:
+    """
+    Return the timestamp of a path or 0, if it does not exist.
+
+    :param path: Path to the file or directory of interest
+    :return: File or directory timestamp
+    """
+    try:
+        timestamp = path.stat().st_mtime
+    except FileNotFoundError:
+        timestamp = 0
+    return timestamp
+
+
+def get_latest_partial_parse(dbt_project_path: Path, cache_dir: Path) -> Path | None:
+    """
+    Return the path to the latest partial parse file, if defined.
+
+    :param dbt_project_path: Original dbt project path
+    :param cache_dir: Path to the Cosmos project cache directory
+    :return: Either return the Path to the latest partial parse file, or None.
+    """
+    project_partial_parse_path = get_partial_parse_path(dbt_project_path)
+    cosmos_cached_partial_parse_filepath = get_partial_parse_path(cache_dir)
+
+    age_project_partial_parse = get_timestamp(project_partial_parse_path)
+    age_cosmos_cached_partial_parse_filepath = get_timestamp(cosmos_cached_partial_parse_filepath)
+
+    if age_project_partial_parse and age_cosmos_cached_partial_parse_filepath:
+        if age_project_partial_parse > age_cosmos_cached_partial_parse_filepath:
+            return project_partial_parse_path
+        else:
+            return cosmos_cached_partial_parse_filepath
+    elif age_project_partial_parse:
+        return project_partial_parse_path
+    elif age_cosmos_cached_partial_parse_filepath:
```

Comment:
I am willing to understand when it can happen that the original project does not have the partial parse file but our cache can have it? Is it that the original project will have a parse file if a user has given in their project and otherwise no. And that when dbt parses the directory(with partial parsing on), it creates this partial parse in the project tmp dir and from there we copy it to our cache?

Reply:
That's a good question, @pankajkoti !

Cosmos users are not guaranteed to make the `partial_parse.msgpack` available to the Airflow scheduler/worker nodes.

This file is just generated after a user runs a dbt command, parses the project, and it is stored in the `target` folder.

Since Cosmos never runs `dbt` commands from within the original project folder, unless the user explicitly had made this file part of their deployment, the original project will not contain it.